spec/abstract/KHeap_A.thy

```diff
@@ -19,14 +19,20 @@ for kernel objects.\<close>
 
 section "General Object Access"
 
+text \<open>The kernel reader option monad\<close>
+type_synonym ('a,'z) r_monad = "('z state, 'a) lookup"
+
+definition
+  read_object :: "obj_ref \<Rightarrow> (kernel_object,'z::state_ext) r_monad"
+where
+  "read_object \<equiv> swp kheap"
+(*  "read_object ptr s = kheap s ptr" *)
+
 definition
   get_object :: "obj_ref \<Rightarrow> (kernel_object,'z::state_ext) s_monad"
 where
-  "get_object ptr \<equiv> do
-     kh \<leftarrow> gets kheap;
-     assert (kh ptr \<noteq> None);
-     return $ the $ kh ptr
-   od"
+  "get_object ptr \<equiv> gets_the $ read_object ptr"
+(* this is equivalent to "get_object ptr \<equiv> gets_map kheap ptr" *)
```

Comment:
It'd be nice to state the comment as a lemma instead, i.e. `lemma get_object_gets_map: "get_object = gets_map kheap"`